File: src/main/java/com/puhovin/intellijplugin/twm/settingsmanager/SettingsManager.java
package com.puhovin.intellijplugin.twm.settingsmanager;

import com.intellij.openapi.components.PersistentStateComponent;
import com.puhovin.intellijplugin.twm.model.ToolWindowPreference;
import com.puhovin.intellijplugin.twm.model.ToolWindowPreferenceStore;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

public interface SettingsManager extends PersistentStateComponent<ToolWindowPreferenceStore> {

    @Override
    @NotNull ToolWindowPreferenceStore getState();

    @Override
    void loadState(@NotNull ToolWindowPreferenceStore state);

    void applyPreferences(@NotNull Map<String, ToolWindowPreference> preferences);

    Map<String, ToolWindowPreference> getDefaultPreferences();

    void setDefaultPreferences(@NotNull Map<String, ToolWindowPreference> preferences);

    ToolWindowPreference getDefaultAvailabilityToolWindow(String id);
}

File: src/main/kotlin/com/puhovin/intellijplugin/twm/settingsmanager/SettingsManager.kt
package com.puhovin.intellijplugin.twm.settingsmanager

import com.intellij.openapi.components.PersistentStateComponent
import com.puhovin.intellijplugin.twm.model.ToolWindowPreference
import com.puhovin.intellijplugin.twm.model.ToolWindowPreferenceStore

interface SettingsManager : PersistentStateComponent<ToolWindowPreferenceStore> {

    override fun getState(): ToolWindowPreferenceStore

    override fun loadState(state: ToolWindowPreferenceStore)

    fun getPreferences(): Map<String, ToolWindowPreference>

    fun setPreferences(preferences: Map<String, ToolWindowPreference?>)

    fun getDefaultPreferences(): Map<String, ToolWindowPreference>

    fun setDefaultPreferences(preferences: Map<String, ToolWindowPreference>)

    fun getDefaultAvailabilityToolWindow(id: String?): ToolWindowPreference?
}